### PR TITLE
Replace custom Future implementations by CompletableFuture

### PR DESCRIPTION
**Pull request by ywelsch**
This PR replaces our custom future implementations by Java 8's `CompletableFuture`, which results in less custom code to maintain (see PR stats), and enables the use of more powerful async programming features across our codebase. Furthermore, it fixes an issue with `CompletableFuture`, where it's not properly bubbling up Errors to the uncaught exception handler.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java [hunk 1]
```diff
@@ -59,68 +54,22 @@ protected PlainListenableActionFuture() {}
...
-        Object listeners = this.listeners;
-        if (listeners != null) {
-            if (listeners instanceof List) {
-                List list = (List) listeners;
-                for (Object listener : list) {
-                    executeListener((ActionListener<T>) listener);
-                }
+        whenComplete((val, throwable) -> {
+            if (throwable == null) {
+                listener.onResponse(val);
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java]: old code caught exceptions here and routed them to listener.onFailure()

File: server/src/main/java/org/elasticsearch/common/util/concurrent/BaseFuture.java [hunk 2]
```diff
@@ -19,324 +19,46 @@
...
     @Override
     public V get(long timeout, TimeUnit unit) throws InterruptedException,
             TimeoutException, ExecutionException {
-        assert timeout <= 0 ||
-            (Transports.assertNotTransportThread(BLOCKING_OP_REASON) &&
-                ThreadPool.assertNotScheduleThread(BLOCKING_OP_REASON) &&
-                ClusterApplierService.assertNotClusterStateUpdateThread(BLOCKING_OP_REASON) &&
-                MasterService.assertNotMasterUpdateThread(BLOCKING_OP_REASON));
-        return sync.get(unit.toNanos(timeout));
+        assert timeout <= 0 || blockingAllowed();
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/BaseFuture.java]: 💯

File: server/src/main/java/org/elasticsearch/common/util/concurrent/ListenableFuture.java [hunk 3]
```diff
@@ -38,76 +35,35 @@
...
-                    run = false;
-                }
-            }
-
-            if (run) {
-                // run the callback directly, we don't hold the lock and don't need to fork!
-                notifyListener(listener, EsExecutors.newDirectExecutorService());
+        whenCompleteAsync((val, throwable) -> {
+            if (throwable == null) {
+                listener.onResponse(val);
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/ListenableFuture.java]: same comment about exception handling.

**Review by bleskes — CHANGES_REQUESTED** [review]
Looks awesome. I left minor comments.

File: server/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java (hunk 1 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java]: Good catch. I'm not a fan of having this anti-pattern in library code but I'm still making this change for BWC purposes. In the future (pun intended) we can also think about adding these methods directly to BaseFuture as every Future has now listening capabilities.

**Comment by tbrooks8** [comment]
I have a comment. This is similar to work I did in `org.elasticsearch.common.concurrent.CompletableContext`. One thing that I have noticed recently is that `CompletableFuture` catches `Throwable` that occurs in listeners. So:

```
CompletableContext<Object> context = new CompletableContext<>();
context.addListener((o, e) -> {
        assert false : "Should not fail";
});
context.complete(null);
```

That assertion is swallowed and never seen again. This is because when the listener is executed, `CompletableFuture` catches it and completes the other `CompletableFuture` that was created when the listener was attached. You have to manually interrogate that future to find it.

```

    final boolean uniWhenComplete(Object r,
                                  BiConsumer<? super T,? super Throwable> f,
                                  UniWhenComplete<T> c) {
        T t; Throwable x = null;
        if (result == null) {
            try {
                if (c != null && !c.claim())
                    return false;
                if (r instanceof AltResult) {
                    x = ((AltResult)r).ex;
                    t = null;
                } else {
                    @SuppressWarnings("unchecked") T tr = (T) r;
                    t = tr;
                }
                f.accept(t, x);
                if (x == null) {
                    internalComplete(r);
                    return true;
                }
            } catch (Throwable ex) {
                if (x == null)
                    x = ex;
                else if (x != ex)
                    x.addSuppressed(ex);
            }
            completeThrowable(x, r);
        }
        return true;
    }
```

I had not decided how I was going to fix that issue. But I was thinking of wrapping attached listeners, catching `Error` and doing the throw it to another thread thing.

I just thought I would mention that here as I assume it will impact this work and it was causing me issues for assertions to disappear.

**Comment by jasontedor** [comment]
Note that we handled a similar problem in #28667.

**Comment by tbrooks8** [comment]
The recent updates seem like they will work. The only downside seems to be that all the `CompletionStage` Apis (like `whenComplete`) take `Throwable`.  People that use the `BaseFuture`  will still have to deal with `Throwable`. But I think that is unavoidable without essentially creating our own `CompletionStage` interface that does not take `Throwable`.

**Comment by ywelsch** [comment]
I have reworked this PR so that the CompletableFutures now properly bubble up Errors to the uncaught exception handler.

> People that use the BaseFuture will still have to deal with Throwable.

yes, but only if they make use of the methods on CompletionStage that require to do so. We can also provide convenient alternatives on BaseFuture if need be.

As follow-ups, I would like to:
- rename BaseFuture to something like just `ESFuture`
- remove some of the subclasses of BaseFuture, as the functionality is now directly available in the super class (e.g. ListenableFuture).

**Pull request closed without merging**